File: Java/FactoryMethod/GuitarFactory/src/com/ericgtkb/GuitarFactory.java
package com.ericgtkb;

public abstract class GuitarFactory {
    public Guitar getGuitar(String type) {
        Guitar guitar = buildGuitar(type);

        guitar.assemble();
        guitar.stringTheGuitar();

        return guitar;
    }

    // The actual factory method
    protected abstract Guitar buildGuitar(String type);
}

class GibsonFactory extends GuitarFactory {
    @Override
    protected Guitar buildGuitar(String type) {
        if (type.equals("Les Paul")) {
            return new GibsonLesPaul();
        } else if (type.equals("SG")) {
            return new GibsonSG();
        } else {
            return null;
        }
    }
}


class FenderFactory extends GuitarFactory {
    @Override
    protected Guitar buildGuitar(String type) {
        if (type.equals("Stratocaster")) {
            return new FenderStratocaster();
        } else if (type.equals("Telecaster")) {
            return new FenderTelecaster();
        } else {
            return null;
        }
    }
}

File: C++/FactoryMethod/GuitarFactory/guitar_factory.cc

#include "guitar_factory.h"

std::unique_ptr<Guitar> GuitarFactory::get_guitar(const std::string& type) const {
    std::unique_ptr<Guitar> guitar {build_guitar(type)};
    guitar->assemble();
    guitar->string_the_guitar();

    return std::move(guitar);
}


std::unique_ptr<Guitar> GibsonFactory::build_guitar(const std::string &type) const {
    if (type == "Les Paul") {
        return std::make_unique<GibsonLesPaul>();
    } else if (type == "SG") {
        return std::make_unique<GibsonSG>();
    } else {
        return nullptr;
    }
}


std::unique_ptr<Guitar> FenderFactory::build_guitar(const std::string &type) const {
    if (type == "Stratocaster") {
        return std::make_unique<FenderStratocaster>();
    } else if (type == "Telecaster") {
        return std::make_unique<FenderTelecaster>();
    } else {
        return nullptr;
    }
}
